ocaml/typing/ctype.ml

```diff
@@ -2195,137 +2125,107 @@ let type_jkind_sub env ty jkind =
+             always right.  Suppose we had [type ('a : wprd) t = 'a] and we were
+             checking ['a t] against [value]. Then it would be enough for [word]
+             and [value] to have an intersection, not for one to be a subjkind
+             of another. But getting this reporting correct would require
+             arbitrary amounts of expansion and looking through [@@unboxed]
+             types. So we don't, settling for the slightly worse error
+             message. *)
+          Error (Jkind.Violation.of_ (Not_a_subjkind (ty's_jkind, jkind)))
+       | Has_intersection ->
+          match get_desc ty with
+          | Tconstr _ ->
+             if not expanded
+             then
+               let ty = expand_head_opt env ty in
+               loop ~fuel ~expanded:true ty (estimate_type_jkind env ty) jkind
+             else
+               begin match unbox_once env ty with
+               | Missing path -> Error (Jkind.Violation.of_ ~missing_cmi:path
+                                          (Not_a_subjkind (ty's_jkind, jkind)))
+               | Final_result ->
+                  Error (Jkind.Violation.of_ (Not_a_subjkind (ty's_jkind, jkind)))
+               | Stepped ty ->
+                  loop ~fuel:(fuel - 1) ~expanded:false ty
+                    (estimate_type_jkind env ty) jkind
+               end
+          | Tunboxed_tuple ltys ->
+             let num_components = List.length ltys in
+             begin match Jkind.decompose_product ty's_jkind,
+                         Jkind.decompose_product jkind with
+             | Some ty's_jkinds, Some jkinds
+                  when List.length ty's_jkinds = num_components
+                       && List.length jkinds = num_components ->
+                  (* Note: here we "duplicate" the fuel, which may seem like
+                     cheating. Fuel counts expansions, and its purpose is to
+                     guard against infinitely expanding a recursive type. In a
+                     wide product, we many need to expand many types shallowly,
+                     and that's fine. *)
+                let results =
+                  Misc.Stdlib.List.map3
+                    (fun (_, ty) -> loop ~fuel ~expanded:false ty)
+                    ltys ty's_jkinds jkinds
+                in
+                let module Monad_result = Misc.Stdlib.Monad.Make2(struct
+                                              include Result
+                                              let return = ok
+                                            end) in
+                Monad_result.all_unit results
+             | _ -> Misc.fatal_error "unboxed tuple jkinds don't line up"
+             end
+          | _ -> Error (Jkind.Violation.of_ (Not_a_subjkind (ty's_jkind, jkind)))
+  in
+  loop ~fuel:100 ~expanded:false ty (estimate_type_jkind env ty) jkind
 
-let constrain_type_jkind ~fixed env ty jkind =
-  (* An optimization to avoid doing any work if we're checking against
```

Comment:
Previously, we still did expansion even when the `jkind` was `any`. Now, we won't. But upon looking at the new implementation, the path to deciding that `any` is fine is a little longer than it could be, especially if there is an unfixed tyvar. I'll put this back.